# Exploring eBay Car Sales Data

In this project, I will be working with a dataset of used cars from *eBay Kleinanzeigen*, a classified section of the German eBay website. 

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np
#tried reading data with specifying encoding, failed
#next try Latin-1
autos = pd.read_csv("autos.csv",encoding = "Latin-1")

In [2]:
#exploring features of jupyter, first few & last few values
#of our data
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

After importing the data, I first want to look at the column headings to become more familiar with what I am working with.  It is apparent that the dataset has 50,000 entries, although it appears that some columns contain null values.  

Comparing the information shown above with the first few rows of the dataset below, I notice a few things that may need cleaned up.  For instance, the 'price' column contains a '$' and is stored as an object type.  Similarly, the 'odometer' column can be converted to a integer type.  Some of the columns are also listed in German, which should be fairly easy to convert to English later on.

Lastly, the column names use camelcase, instead of snakecase, which is my preferred method.  So I will convert those next

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
#Copy above array and make changes
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']
autos.columns = new_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Above, I converted the column names from camelcase to snakecase by copying the current column names and manually making changes.  I did this because I believe it increases the readability of the data.

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [8]:
autos[["seller","offer_type","price","odometer","num_pictures"]].head(20)

seller offer_type    price   odometer  num_pictures
0   privat    Angebot   $5,000  150,000km             0
1   privat    Angebot   $8,500  150,000km             0
2   privat    Angebot   $8,990   70,000km             0
3   privat    Angebot   $4,350   70,000km             0
4   privat    Angebot   $1,350  150,000km             0
5   privat    Angebot   $7,900  150,000km             0
6   privat    Angebot     $300  150,000km             0
7   privat    Angebot   $1,990  150,000km             0
8   privat    Angebot     $250  150,000km             0
9   privat    Angebot     $590  150,000km             0
10  privat    Angebot     $999  150,000km             0
11  privat    Angebot     $350  150,000km             0
12  privat    Angebot   $5,299   50,000km             0
13  privat    Angebot   $1,350  150,000km             0
14  privat    Angebot   $3,999  150,000km             0
15  privat    Angebot  $18,900   80,000km             0
16  privat    Angebot     $350  150,000km             0
17  privat    Angebot   $5,500  150,000km             0
18  privat    Angebot     $300  150,000km             0
19  privat    Angebot   $4,150  150,000km             0

In [9]:
autos["seller"].unique()

array(['privat', 'gewerblich'], dtype=object)

In [10]:
autos["offer_type"].unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [11]:
autos[autos["seller"]=="privat"]["offer_type"].value_counts()

Angebot    49998
Gesuch         1
Name: offer_type, dtype: int64

In [12]:
autos["num_pictures"].value_counts()

0    50000
Name: num_pictures, dtype: int64

Investigating the data a little bit more, I noticed a few things.  For one, both the *seller* and *offer_type* columns only have two unique values, with each accounting for only one entry out of 50,000 for each.  Which means that 49,998 of the entries have "privat" and "Angebot" entered in each respective column, making one of these columns redundant.

Another thing to note is that the *price* and *odometer* columns are both stored as text with units in each entry, when they could be stored as numbers which would allow for further analysis.

The last thing I want to point out is that the *num_pictures* column is equal to 0 for every single entry.  Thus we probably do not need it in our dataset.

The first task I will tackle is converting the *price* and *odometer* columns to numeric types.

In [13]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(int)
autos.rename({"price":"price_dollar"},axis=1, inplace=True)

autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].astype(int)
autos.rename({"odometer":"odometer_km"},axis=1, inplace=True)

In [14]:
autos[["price_dollar","odometer_km"]].head(5)

price_dollar  odometer_km
0          5000       150000
1          8500       150000
2          8990        70000
3          4350        70000
4          1350       150000

In [15]:
autos["price_dollar"].unique().shape

(2357,)

In [16]:
autos["price_dollar"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64

In [17]:
autos["price_dollar"].value_counts().sort_index(ascending=True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price_dollar, dtype: int64

In [18]:
autos["price_dollar"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_dollar, dtype: int64

In [19]:
autos[autos["price_dollar"] > 150000]["price_dollar"].value_counts().sort_index(ascending=False)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price_dollar, dtype: int64

Looking at the price data, a couple red flags pop up.  The first being that there were 1421 cars sold for \$0, which does not make sense because they were sold on eBay where the starting bid is $1.  On the opposite end of the spectrum, some of the prices were suspect, including one car that has a listed price of almost \$100 million.  

Looking at the prices that are over \$150 thousand, it appears that prices steadily increase until they get to \$350 thousand, at which point they sharply increase to almost \$1 million.  I believe this would be a good point to remove outliers because the data above this point appears to be unreliable.

In [20]:
autos = autos[autos["price_dollar"].between(1,351000)]
autos["price_dollar"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_dollar, dtype: float64

In [21]:
50000-48565

1435

A total of 1435 entries were removed from the dataset, but I believe that this makes our data more accurate.

# Exploring the Date Columns

There are a couple of different columns with date information in them:
1. *date_crawled*
2. *registration_month*
3. *registration_year*
4. *ad_created*
5. *last_seen*

Both *registration_month* and *registration_year* are stored as numeric values, but the rest are stored as strings, so I will need to convert these to a numerical representation so they can be understood quantitatively.

In [22]:
autos[["date_crawled","ad_created","last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [23]:
(autos["date_crawled"]
                    .str[:10]
                    .value_counts(normalize=True, dropna=False)
                    .sort_index())

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [24]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values(ascending=False))

2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

The crawler also recorded what date the listing was last seen, which is likely when the car was sold.

In [25]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

One thing to note with this information is that there is a disproportionate amount of 'last seen' values on the last couple of days, which could either mean a very large spike in sales, or that the crawling period was ending, which is much more likely.  

In [26]:
(autos["ad_created"]
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

This is a total of 76 dates on which ads were created dating back to June of 2015, but the vast majority of posts come in March and April of 2016.  

In [27]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There appears to be some incorrect data in the *registration_year* column, as the minimum value is 1000 and the maximum value is 9999, neither of which make sense for registration years of a vehicle.

In [28]:
print(autos["registration_year"]
    .value_counts()
    .sort_index()
    .head(10))
print('\n')
print(autos["registration_year"]
    .value_counts()
    .sort_index(ascending=False)
    .head(15))

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
Name: registration_year, dtype: int64


9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
Name: registration_year, dtype: int64


Determining which data is inaccurate is fairly easy for the higher end of registration years.  Since data was collected in 2016, any vehicle with a registration year greater than that value is inaccurate.  

Dealing with the lower values is more complicated.  It is easy to conclude that anything with the values of 1000-1800 are inaccurate, but determining a cutoff year is tricky.  Realistically, it could be somewhere in the first few decades of the 1900s.  

Let's see what percent of data remains if I set the range to be 1900-2016

In [29]:
autos["registration_year"].between(1900,2016).sum()/autos.shape[0]

0.961206630289303

This eliminates less than 4% of the data, so we will remove the rows outside of this range.

In [30]:
autos = autos[autos["registration_year"].between(1900,2016)]
print(autos["registration_year"].value_counts(normalize=True, dropna = False)
    .sort_index()
    .head(10))
print('\n')
print(autos["registration_year"].value_counts(normalize=True, dropna = False)
    .sort_index(ascending=False)
    .head(20))

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
Name: registration_year, dtype: float64


2016    0.026135
2015    0.008397
2014    0.014203
2013    0.017202
2012    0.028063
2011    0.034768
2010    0.034040
2009    0.044665
2008    0.047450
2007    0.048778
2006    0.057197
2005    0.062895
2004    0.057904
2003    0.057818
2002    0.053255
2001    0.056468
2000    0.067608
1999    0.062060
1998    0.050620
1997    0.041794
Name: registration_year, dtype: float64


It would appear the the majority of the cars listed were registered in the last 20 years.

# Exploring Price by Brand

In [31]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

There are a total of 40 brands of cars across all of the listings, with the majority of them being German car brands. Now I want to analyze the price of the listings by brand, but there are lots of brand that don't have a significant percentage of listings.  So I will limit the scope of this analysis to brands representing more than 5% of total listings 

In [32]:
popular_brands = autos["brand"].value_counts(normalize=True)
popular_brands = popular_brands[popular_brands > 0.05].index
print(popular_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [33]:
brand_mean_prices = {}
for brand in popular_brands:
    car_brands = autos[autos["brand"]==brand]
    mean_price = car_brands["price_dollar"].mean()
    brand_mean_prices[brand] = mean_price
    
brand_mean_prices

{'audi': 9336.687453600594,
 'bmw': 8332.820517811953,
 'ford': 3749.4695065890287,
 'mercedes_benz': 8628.450366422385,
 'opel': 2975.2419354838707,
 'volkswagen': 5402.410261610221}

Out of the brands that accounted for more than 5% of sales, Audi had the highest average selling price at around \$9300.  BMW and Mercedes-Benz also had relatively high mean prices, at over \$8000 apiece.  Logically, this makes sense because all three of these brands are considered to be luxury here in the United States which would increase their value.

Ford and Opel both have much cheaper average selling prices, with both being less than \$4000.  Volkswagen is right in the middle of all of these brands, with an average selling price of around \$5400.

# Exploring Car Mileage

In [34]:
#Converting dictionary into a series
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

audi             9336.687454
bmw              8332.820518
ford             3749.469507
mercedes_benz    8628.450366
opel             2975.241935
volkswagen       5402.410262
dtype: float64


In [35]:
new_df = pd.DataFrame(bmp_series, columns=["mean_price"])
print(new_df)

                mean_price
audi           9336.687454
bmw            8332.820518
ford           3749.469507
mercedes_benz  8628.450366
opel           2975.241935
volkswagen     5402.410262


In [36]:
brand_mean_mileage = {}
for brand in popular_brands:
    car_brands = autos[autos["brand"] == brand]
    mean_mileage = car_brands["odometer_km"].mean()
    brand_mean_mileage[brand] = mean_mileage
brand_mean_mileage

{'audi': 129157.38678544914,
 'bmw': 132572.51313996495,
 'ford': 124266.01287159056,
 'mercedes_benz': 130788.36331334666,
 'opel': 129310.0358422939,
 'volkswagen': 128707.15879132022}

In [37]:
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_price = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [38]:
brand_summary = pd.DataFrame(mean_mileage, columns=["mean_mileage"])

In [39]:
brand_summary["mean_price"] = mean_price
brand_summary

mean_mileage   mean_price
bmw            132572.513140  8332.820518
mercedes_benz  130788.363313  8628.450366
opel           129310.035842  2975.241935
audi           129157.386785  9336.687454
volkswagen     128707.158791  5402.410262
ford           124266.012872  3749.469507

Average mileages for each brand do not very too much.  BMW, which has the highest mean mileage amongst the brands, is only about 6% higher than the lowest mean mileage, which belongs to Ford.  Meanwhile, BMW's mean price is more than 120% more than Fords, which leads me to believe that there is not a strong correlation between the mileage of the car and the selling price.

# Next Steps

In [40]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_dollar  ab_test vehicle_type  registration_year  \
0  privat    Angebot          5000  control          bus               2004   
1  privat    Angebot          8500  control    limousine               1997   
2  privat    Angebot          8990     test    limousine               2009   
3  privat    Angebot          4350  control   kleinwagen               2007   
4  privat    Angebot          1350     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

As you can see in the dataframe, some single words columns are stored in German, and I plan to replace these with their English counterpart.

In [41]:
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In the vehicle type column, there are a few words that will need to be translated, which I have done using Google Translate.  Some may not be perfect translations.  For instance, "kleinwagen" translates to "small car", which I am interpreting to mean "compact car".  

In [42]:
autos.loc[autos["vehicle_type"]== "kleinwagen", "vehicle_type"] = "compact_car"
autos.loc[autos["vehicle_type"]== "kombi", "vehicle_type"] = "station_wagon"
autos.loc[autos["vehicle_type"]== "cabrio", "vehicle_type"] = "convertible"
autos.loc[autos["vehicle_type"]== "andere", "vehicle_type"] = "other"

In [43]:
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'compact_car', 'station_wagon', nan, 'coupe',
       'suv', 'convertible', 'other'], dtype=object)

In [44]:
autos["seller"].unique()

array(['privat', 'gewerblich'], dtype=object)

In [45]:
autos.loc[autos["seller"]=="privat", "seller"] = "private"
autos.loc[autos["seller"]=="gewerblich", "seller"] = "commercial"
autos["seller"].unique()

array(['private', 'commercial'], dtype=object)

In [46]:
autos["offer_type"].unique()

array(['Angebot'], dtype=object)

This column only has one unique value, so it does not need to be included in the dataset

In [47]:
del autos["offer_type"]

In [48]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    seller  price_dollar  ab_test   vehicle_type  registration_year  \
0  private          5000  control            bus               2004   
1  private          8500  control      limousine               1997   
2  private          8990     test      limousine               2009   
3  private          4350  control    compact_car               2007   
4  private          1350     test  station_wagon               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [49]:
autos["gearbox"].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [50]:
autos.loc[autos["gearbox"]=="manuell", "gearbox"] = "manual"
autos.loc[autos["gearbox"]=="automatik", "gearbox"] = "automatic"

In [51]:
autos["fuel_type"].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [52]:
autos.loc[autos["fuel_type"]=="benzin", "fuel_type"] = "gasoline"
autos.loc[autos["fuel_type"]=="elektro", "fuel_type"] = "electric"
autos.loc[autos["fuel_type"]=="andere", "fuel_type"] = "other"

In [53]:
autos["unrepaired_damage"].unique()

array(['nein', nan, 'ja'], dtype=object)

In [54]:
autos.loc[autos["unrepaired_damage"]=="nein", "unrepaired_damage"] = "no"
autos.loc[autos["unrepaired_damage"]=="ja", "unrepaired_damage"] = "yes"

In [55]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    seller  price_dollar  ab_test   vehicle_type  registration_year  \
0  private          5000  control            bus               2004   
1  private          8500  control      limousine               1997   
2  private          8990     test      limousine               2009   
3  private          4350  control    compact_car               2007   
4  private          1350     test  station_wagon               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0     manual       158  andere       150000                   3       lpg   
1  automatic       286     7er       150000                   6  gasoline   
2     manual       102    golf        70000                   7  gasoline   
3  automatic        71  fortwo        70000                   6  gasoline   
4     manual         0   focus       150000                   7  gasoline   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot                no  2016-03-26 00:00:00             0   
1         bmw                no  2016-04-04 00:00:00             0   
2  volkswagen                no  2016-03-26 00:00:00             0   
3       smart                no  2016-03-12 00:00:00             0   
4        ford                no  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

It looks like I have cleaned up all of the columns that contained German words, except for the the *model* column, which contains too many values to translate them all.

In [56]:
autos["model"].unique().shape

(245,)

Next I will be converting date/time columns to be more consistent. Instead of including year/month/day/time, it will be stored as YYYYMMDD as an integer

In [57]:
#Enter in column name as "column_name"
def string_date_to_int(column):
    autos[column] = autos[column].str[:10]
    autos[column] = autos[column].str.replace("-","")
    autos[column] = autos[column].astype(int)
    return 

In [58]:
string_date_to_int("date_crawled")

In [59]:
string_date_to_int("ad_created")

In [60]:
string_date_to_int("last_seen")

In [61]:
autos.head()

date_crawled                                               name   seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2      20160326                         Volkswagen_Golf_1.6_United  private   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

   price_dollar  ab_test   vehicle_type  registration_year    gearbox  \
0          5000  control            bus               2004     manual   
1          8500  control      limousine               1997  automatic   
2          8990     test      limousine               2009     manual   
3          4350  control    compact_car               2007  automatic   
4          1350     test  station_wagon               2003     manual   

   power_PS   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6  gasoline         bmw   
2       102    golf        70000                   7  gasoline  volkswagen   
3        71  fortwo        70000                   6  gasoline       smart   
4         0   focus       150000                   7  gasoline        ford   

  unrepaired_damage  ad_created  num_pictures  postal_code  last_seen  
0                no    20160326             0        79588   20160406  
1                no    20160404             0        71034   20160406  
2                no    20160326             0        35394   20160406  
3                no    20160312             0        33729   20160315  
4                no    20160401             0        39218   20160401

As you can see, these date columns are now much cleaner and easier to read than before.  The exact time included in each column was not important, so I narrowed the column down to just included the year, month, and day.

Next I will take a look at the most common brand/model combinations.  For this, I will need to sort the cars first by brand, and then by model.  

In [62]:
autos["model"].groupby(autos["brand"]).value_counts().sort_values(
    ascending=False).head(10)

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1592
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
Name: model, dtype: int64

This is a list of the most popular models of car listed, grouped by the car's brand.  The Volkswagen Golf is by far the most popular car listed, with over 3700 (but just over 1% of the total) listings.  The top 10 is dominated by Volkswagen, BMW, Opel, Mercedes-Benz, and Audi.  If you recall from earlier, these were also our top 5 overall car brands in terms of listings, so it would make sense that their models appear more.

# How Mileage Affects Price

In [63]:
autos["odometer_km"].describe()

count     46681.000000
mean     125586.855466
std       39852.528628
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [79]:
price_less_than_100km = autos.loc[autos["odometer_km"] < 100000,"price_dollar"]
price_100_to_125km = autos[autos["odometer_km"].between(100000,125000)]["price_dollar"]
price_125_to_150km = autos[autos["odometer_km"].between(125000,150000)]["price_dollar"]
price_150km = autos.loc[autos["odometer_km"] >= 150000,"price_dollar"]

avg_prices = {}
avg_prices["less than 100km"] = int(price_less_than_100km.mean())
avg_prices["100 to 125km"] = int(price_100_to_125km.mean())
avg_prices["125 to 150km"] = int(price_125_to_150km.mean())
avg_prices["150km"] = int(price_150km.mean())

avg_prices

{'100 to 125km': 6785,
 '125 to 150km': 4107,
 '150km': 3767,
 'less than 100km': 12268}

As you can see from this list, the price of the vehicles appears to decrease as the mileage increases.  Vehicles with less than 100,000 km sold for an average of over \$12,000.  That number drops to around \$3800 for vehicles 150,000km (which was the maximum odometer number seen in our data).  This number may be skewed though, as over 30,000 vehicles had an odometer reading of 150,000.  This likely means that the options when listing the vehicle were limited, and that the actual mile count is higher than 150,000.  Nevertheless, you can still see how the prices trend downward as mileage increases.

In [65]:
autos["odometer_km"].value_counts().head(1)

150000    30085
Name: odometer_km, dtype: int64

# How Damage Impacts Price

The last thing that I want to take a look at in this project is how the average prices compare for damaged cars and their non-damaged counterparts.

In [66]:
autos["unrepaired_damage"].value_counts()

no     33834
yes     4540
Name: unrepaired_damage, dtype: int64

In [83]:
price_damage = autos.loc[autos["unrepaired_damage"]=="yes", "price_dollar"]
price_no_damage = autos.loc[autos["unrepaired_damage"]=="no", "price_dollar"]

print('Damaged cars summary')
print(price_damage.describe())
print('\n')
print('Non-damaged cars summary')
print(price_no_damage.describe())


Damaged cars summary
count     4540.000000
mean      2241.146035
std       3563.276478
min          1.000000
25%        500.000000
50%       1000.000000
75%       2500.000000
max      44200.000000
Name: price_dollar, dtype: float64


Non-damaged cars summary
count     33834.000000
mean       7164.033103
std       10078.475478
min           1.000000
25%        1800.000000
50%        4150.000000
75%        9000.000000
max      350000.000000
Name: price_dollar, dtype: float64


The average price of damaged cars is significantly lower than the average price of non-damaged cars, which is to be expected.  